File: python/examples/convert_sample.py
"""Convert the bundled sample WAV file to μ-law 8KHz output."""

from __future__ import annotations

import argparse
import math
import wave
from pathlib import Path

import numpy as np
from scipy import signal
from scipy.io import wavfile

from audio_converter import AudioStreamConverter, MuLawStreamDecoder

DATA_DIR = Path(__file__).resolve().parent / "data"
DEFAULT_INPUT = DATA_DIR / "sample_pcm16_24khz.wav"
DEFAULT_ULAW = DATA_DIR / "down_sample_8khz.ulaw"
DEFAULT_DECODED_8K = DATA_DIR / "down_sample_8khz.wav"
DEFAULT_UPSAMPLED = DATA_DIR / "up_sample_pcm16_24khz_decoded.wav"


def to_mono_pcm16_24khz(path: Path) -> np.ndarray:
    """Load WAV audio and return mono PCM16 24KHz samples.

    The bundled source is a 6-channel capture; we keep the first channel to
    preserve the original speaker while avoiding phase cancellation that
    occurs when averaging channels containing decorrelated ambience.
    """
    rate, data = wavfile.read(str(path))
    array = data.astype(np.float32, copy=False)

    if array.ndim > 1:
        array = array[:, 0]

    peak = float(np.max(np.abs(array))) or 1.0
    array = array / peak

    target_rate = 24000
    if rate != target_rate:
        gcd = math.gcd(rate, target_rate)
        up = target_rate // gcd
        down = rate // gcd
        array = signal.resample_poly(array, up, down).astype(np.float32, copy=False)

    pcm16 = np.clip(array, -1.0, 1.0)
    return (pcm16 * 32767).astype(np.int16)


def convert_wav_to_ulaw(input_wav: Path, output_ulaw: Path) -> None:
    """Convert the bundled WAV file to μ-law 8KHz using the converter."""
    pcm16 = to_mono_pcm16_24khz(input_wav)
    converter = AudioStreamConverter()

    with output_ulaw.open("wb") as outfile:
        for start in range(0, len(pcm16), converter.chunk_size):
            end = start + converter.chunk_size
            chunk = converter.convert_chunk(pcm16[start:end].tobytes())
            if chunk:
                outfile.write(chunk)

        final = converter.flush()
        if final:
            outfile.write(final)


def decode_ulaw_to_wav(
    input_ulaw: Path,
    output_wav: Path,
    *,
    input_rate: int = 8000,
    output_rate: int = 24000,
    apply_smoothing: bool = True,
) -> None:
    """Decode μ-law bytes to PCM16 and persist as a WAV file."""
    decoder = MuLawStreamDecoder(input_rate=input_rate, output_rate=output_rate)
    decoder.reset()

    pcm_chunks: list[np.ndarray] = []
    with input_ulaw.open("rb") as infile:
        chunk_size = decoder.chunk_size
        while chunk := infile.read(chunk_size):
            pcm = decoder.convert_chunk(chunk)
            if pcm:
                pcm_chunks.append(np.frombuffer(pcm, dtype=np.int16))

    final = decoder.flush()
    if final:
        pcm_chunks.append(np.frombuffer(final, dtype=np.int16))

    if not pcm_chunks:
        audio = np.empty(0, dtype=np.int16)
    else:
        audio = np.concatenate(pcm_chunks)

    if apply_smoothing and audio.size and output_rate > input_rate:
        nyquist = output_rate / 2.0
        cutoff = min(0.45 * input_rate, nyquist * 0.95)
        sos = signal.butter(4, cutoff, fs=output_rate, output="sos")
        smoothed = signal.sosfiltfilt(sos, audio.astype(np.float32))
        audio = np.clip(smoothed, -32768, 32767).astype(np.int16)

    with wave.open(str(output_wav), "wb") as wav_out:
        wav_out.setnchannels(1)
        wav_out.setsampwidth(2)
        wav_out.setframerate(decoder.output_rate)
        wav_out.writeframes(audio.tobytes())


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Convert bundled sample WAV to μ-law 8KHz via AudioStreamConverter."
    )
    parser.add_argument(
        "--input",
        type=Path,
        default=DEFAULT_INPUT,
        help=f"Input WAV file (default: {DEFAULT_INPUT})",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=DEFAULT_ULAW,
        help=f"Output μ-law file (default: {DEFAULT_ULAW})",
    )
    parser.add_argument(
        "--decoded-8k",
        type=Path,
        default=DEFAULT_DECODED_8K,
        help=f"Decoded 8KHz WAV output (default: {DEFAULT_DECODED_8K})",
    )
    parser.add_argument(
        "--decoded-24k",
        type=Path,
        default=DEFAULT_UPSAMPLED,
        help=f"Decoded 24KHz WAV output (default: {DEFAULT_UPSAMPLED})",
    )
    return parser


def main() -> None:
    parser = build_parser()
    args = parser.parse_args()
    convert_wav_to_ulaw(args.input, args.output)
    decode_ulaw_to_wav(
        args.output,
        args.decoded_8k,
        input_rate=8000,
        output_rate=8000,
    )
    decode_ulaw_to_wav(
        args.output,
        args.decoded_24k,
        input_rate=8000,
        output_rate=24000,
        apply_smoothing=True,
    )
    print(f"Converted {args.input} → {args.output}")
    print(f"Decoded μ-law to {args.decoded_8k}")
    print(f"Upsampled μ-law to {args.decoded_24k}")


if __name__ == "__main__":
    main()

File: node/src/examples/convertSample.ts
import { readFileSync, writeFileSync } from 'fs';
import { resolve } from 'path';
import { AudioStreamConverter, MuLawStreamDecoder } from '../index';

interface WavInfo {
  sampleRate: number;
  channels: number;
  bitsPerSample: number;
  samples: Int16Array;
}

function parseWav(path: string): WavInfo {
  const buffer = readFileSync(path);
  if (buffer.toString('ascii', 0, 4) !== 'RIFF' || buffer.toString('ascii', 8, 12) !== 'WAVE') {
    throw new Error('Unsupported WAV header');
  }

  let offset = 12;
  let sampleRate = 0;
  let channels = 0;
  let bitsPerSample = 0;
  let dataOffset = 0;
  let dataSize = 0;

  while (offset + 8 <= buffer.length) {
    const chunkId = buffer.toString('ascii', offset, offset + 4);
    const chunkSize = buffer.readUInt32LE(offset + 4);
    const chunkDataStart = offset + 8;

    if (chunkId === 'fmt ') {
      const audioFormat = buffer.readUInt16LE(chunkDataStart);
      if (audioFormat !== 1) {
        throw new Error('Only PCM WAV files are supported');
      }
      channels = buffer.readUInt16LE(chunkDataStart + 2);
      sampleRate = buffer.readUInt32LE(chunkDataStart + 4);
      bitsPerSample = buffer.readUInt16LE(chunkDataStart + 14);
    } else if (chunkId === 'data') {
      dataOffset = chunkDataStart;
      dataSize = chunkSize;
      break;
    }

    offset = chunkDataStart + chunkSize + (chunkSize % 2);
  }

  if (!dataOffset || !sampleRate || bitsPerSample !== 16) {
    throw new Error('Malformed WAV file');
  }

  const totalSamples = dataSize / 2;
  const pcmAll = new Int16Array(totalSamples);
  for (let i = 0; i < totalSamples; i++) {
    pcmAll[i] = buffer.readInt16LE(dataOffset + i * 2);
  }

  let mono: Int16Array;
  if (channels <= 1) {
    mono = pcmAll;
  } else {
    const frames = totalSamples / channels;
    mono = new Int16Array(frames);
    for (let i = 0; i < frames; i++) {
      mono[i] = pcmAll[i * channels];
    }
  }

  return { sampleRate, channels, bitsPerSample, samples: mono };
}

function writeWav(path: string, data: Buffer, sampleRate: number): void {
  const header = Buffer.alloc(44);
  const dataSize = data.length;
  header.write('RIFF', 0);
  header.writeUInt32LE(36 + dataSize, 4);
  header.write('WAVE', 8);
  header.write('fmt ', 12);
  header.writeUInt32LE(16, 16);
  header.writeUInt16LE(1, 20);
  header.writeUInt16LE(1, 22);
  header.writeUInt32LE(sampleRate, 24);
  header.writeUInt32LE(sampleRate * PCM_BYTES_PER_SAMPLE, 28);
  header.writeUInt16LE(PCM_BYTES_PER_SAMPLE, 32);
  header.writeUInt16LE(16, 34);
  header.write('data', 36);
  header.writeUInt32LE(dataSize, 40);
  writeFileSync(path, Buffer.concat([header, data]));
}

const PCM_BYTES_PER_SAMPLE = 2;

function int16ArrayToBytes(data: Int16Array): Buffer {
  const buffer = Buffer.alloc(data.length * 2);
  for (let i = 0; i < data.length; i++) {
    buffer.writeInt16LE(data[i], i * 2);
  }
  return buffer;
}

function decodeAll(decoder: MuLawStreamDecoder, data: Buffer): Buffer {
  const chunk = decoder.chunkSize * 1;
  const buffers: Buffer[] = [];
  for (let offset = 0; offset < data.length; offset += chunk) {
    const converted = decoder.convertChunk(data.subarray(offset, Math.min(offset + chunk, data.length)));
    if (converted) {
      buffers.push(Buffer.isBuffer(converted) ? converted : Buffer.from(converted));
    }
  }
  const finalChunk = decoder.flush();
  if (finalChunk) {
    buffers.push(Buffer.isBuffer(finalChunk) ? finalChunk : Buffer.from(finalChunk));
  }
  return Buffer.concat(buffers);
}

function ensureTwentyFourK(samples: Int16Array, sampleRate: number): Int16Array {
  if (sampleRate !== 24000) {
    console.warn(`Warning: expected 24000 Hz input, received ${sampleRate} Hz. Continuing without resampling.`);
  }
  return samples;
}

function encodeStream(converter: AudioStreamConverter, samples: Int16Array): Buffer {
  const buffers: Buffer[] = [];
  const chunkSize = converter.chunkSize;
  for (let offset = 0; offset < samples.length; offset += chunkSize) {
    const chunk = samples.subarray(offset, Math.min(offset + chunkSize, samples.length));
    const bytes = int16ArrayToBytes(chunk);
    const converted = converter.convertChunk(bytes);
    if (converted) {
      buffers.push(Buffer.isBuffer(converted) ? converted : Buffer.from(converted));
    }
  }
  const finalChunk = converter.flush();
  if (finalChunk) {
    buffers.push(Buffer.isBuffer(finalChunk) ? finalChunk : Buffer.from(finalChunk));
  }
  return Buffer.concat(buffers);
}

function main(): void {
  const dataDir = resolve(__dirname, '../../src/examples/data');
  const inputWav = resolve(dataDir, 'sample_pcm16_24khz.wav');
  const ulawPath = resolve(dataDir, 'down_sample_8khz.ulaw');
  const downWavPath = resolve(dataDir, 'down_sample_8khz.wav');
  const upWavPath = resolve(dataDir, 'up_sample_pcm16_24khz_decoded.wav');

  const wav = parseWav(inputWav);
  const pcm24 = ensureTwentyFourK(wav.samples, wav.sampleRate);

  const converter = new AudioStreamConverter();
  const ulawData = encodeStream(converter, pcm24);
  writeFileSync(ulawPath, ulawData);

  const ulawBytes = readFileSync(ulawPath);
  const decoder8k = new MuLawStreamDecoder(8000, 8000);
  const decoded8k = decodeAll(decoder8k, ulawBytes);
  writeWav(downWavPath, decoded8k, 8000);

  const decoder24k = new MuLawStreamDecoder(8000, 24000);
  const decoded24k = decodeAll(decoder24k, ulawBytes);
  writeWav(upWavPath, decoded24k, 24000);

  console.log(`Converted ${inputWav} → ${ulawPath}`);
  console.log(`Decoded μ-law to ${downWavPath}`);
  console.log(`Upsampled μ-law to ${upWavPath}`);
}

if (require.main === module) {
  main();
}
